packages/vite/src/node/utils.ts

```diff
@@ -1069,6 +1070,74 @@ function backwardCompatibleWorkerPlugins(plugins: any) {
   return []
 }
 
+function deepClone<T>(value: T): T {
+  if (Array.isArray(value)) {
+    return value.map((v) => deepClone(v)) as T
+  }
+  if (isObject(value)) {
+    const cloned: Record<string, any> = {}
+    for (const key in value) {
+      cloned[key] = deepClone(value[key])
+    }
+    return cloned as T
+  }
+  if (typeof value === 'function') {
+    return value as T
+  }
+  return structuredClone(value)
+}
+
+type MaybeFallback<D, V> = undefined extends V ? Exclude<V, undefined> | D : V
+
+type MergeWithDefaultsResult<D, V> =
+  Equal<D, undefined> extends true
+    ? V
+    : D extends Function | Array<any>
+      ? MaybeFallback<D, V>
+      : V extends Function | Array<any>
+        ? MaybeFallback<D, V>
+        : D extends Record<string, any>
+          ? V extends Record<string, any>
+            ? {
+                [K in keyof D | keyof V]: K extends keyof D
+                  ? K extends keyof V
+                    ? MergeWithDefaultsResult<D[K], V[K]>
+                    : D[K]
+                  : K extends keyof V
+                    ? V[K]
+                    : never
+              }
+            : MaybeFallback<D, V>
+          : MaybeFallback<D, V>
+
+export function mergeWithDefaults<
+  D extends Record<string, any>,
+  V extends Record<string, any>,
+>(defaults: D, values: V): MergeWithDefaultsResult<D, V> {
+  // NOTE: we need to clone the value here to avoid mutating the defaults
+  const merged: Record<string, any> = deepClone(defaults)
+  for (const key in values) {
```

Comment:
Since `mergeWithDefaults` is being called recursively, when merging nested objects, wouldn't the object be cloned twice? Should we try to optimize this here? I think we'd have to make the cloning inside of `mergeWithDefaults`, and only clone for keys that are in `defaults` but not in `values`.

Reply:
Good catch! Instead of doing that, I made the `deepClone` to be call only once before merging with defaults (ffb92fdf743026c98edd8830e0868878db36f060). This way, the properties of defaults that would be overridden with the values will be cloned unnecessarily compared to your suggestion, but I think that's negligible.

Reply:
Works for me!